### DISCUSSION TITLE
How do you increase the quality of shadow edges?

### DISCUSSION THREAD

**Discussion by chrisgervang** [visgl/deck.gl#6818] [Q&A] [answered]
#3543 Made an excellent [demo](https://sunlight-to-parks.netlify.app/) of deck.gl sunlight shadows. I've started to use them too, but I've noticed all of the shadow edges look pixelated / aliased. Is there a way to increase the quality of the shadows? Antialiasing?

I'm using deck 8.7 and I am not using any basemap integration in my project, so the answer doesn't need to consider them. 

<img width="577" alt="Screen Shot 2022-04-12 at 9 15 12 AM" src="https://user-images.githubusercontent.com/2461547/163007802-3820c9fd-4ef6-4229-b44e-a2c300d9276a.png">
<img width="767" alt="Screen Shot 2022-04-12 at 9 15 07 AM" src="https://user-images.githubusercontent.com/2461547/163007786-d13b0f9f-8a6a-4b8b-92a0-8360767a8fa7.png">

**Comment by ibgreen** [comment]
FWIW - Reducing shadow noise is a major topic in 3d rendering. There are multiple issues and techniques. I haven't looked at this in a good while so I probably need to refresh my memory but:

- One problem in perspective mode is that different distances from the camera get very different pixel coverage and shadows get pixelated at the edge. Dividing the screen in strips and rendering shadows separately in each strip is a way around that, but fairly involved.
- Another is camera movement can trigger small. but very disturbing flicker in the shadows

There are a lot of great articles on the topic.

**Reply by chrisgervang** [reply]
Is that the same case for FirstPersonView and still cameras as well? 

I'm not really sure where to start to understand how the existing shadows are implemented. I've read their code, but I'm not really sure what to look for when it comes to a root cause of the chunky/pixelated/distorted edge effects.

Is it a matter of some resolution being too low, or something in the shadow render pass? (These are very uninformed questions - really just trying to get my bearings)

Relevant RFCs, but they don't really get into shadow quality:

https://github.com/visgl/deck.gl/tree/master/dev-docs/RFCs/v7.0

https://github.com/visgl/deck.gl/blob/master/dev-docs/RFCs/v7.2/shadow-for-directional-lights.md

**Comment by chrisgervang** [comment, marked as answer]
Increasing `useDevicePixels` helps a lot. I imagine this is because the shadow pass [framebuffer size is determined by it](https://github.com/visgl/deck.gl/blob/1cf1eb0ec0fc83507f6636993321d5fb25b6425d/modules/core/src/passes/shadow-pass.ts#L62-L63), but I haven't confirmed. 

```
<DeckGL
  useDevicePixels={2}
>
</DeckGL>
 ```
Set to `1`
![DPI one](https://user-images.githubusercontent.com/2461547/163055043-f4937939-12c0-450e-b8c6-bb4e99168431.png)
Set to `2`
![DPI two](https://user-images.githubusercontent.com/2461547/163055047-2c3cd88c-c8ca-4c9f-9417-75059887fdb3.png)

I might mark this as the answer if there aren't any other ways to improve shadows right now.

**Reply by chrisgervang** [reply]
Adding the FXAA as a PostprocessingEffect effect also improved shadow quality slightly. I've also noticed different GLTF models at the same position cast different quality shadows, but since I didn't create the models I don't know why this is the case.
